### ISSUE TITLE
Uncaught ZeroDivisionError in SMod

### ISSUE THREAD

**Issue by XiaoyuHe** [angr/angr#704]
# claripy bug:
 situation:
        Test ffmpeg(n3.1.11)  using angr 's undear_constrained_symexecution function at 0x7b2fa0.

 descritpion:
       ZeroDivisionError: long division or modulo by zero

```
Traceback (most recent call last):
  File "constraints.py", line 33, in <module>
    pg.explore()
  File "/root/angr-dev/angr/angr/manager.py", line 917, in explore
    n=n)
  File "/root/angr-dev/angr/angr/manager.py", line 941, in run
    return self.step(n=n, step_func=step_func, until=until_func, stash=stash)
  File "/root/angr-dev/angr/angr/manager.py", line 604, in step
    pg = pg._one_step(stash=stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/manager.py", line 352, in _one_step
    out = hook(pg, stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/exploration_techniques/explorer.py", line 101, in step
    return simgr.step(stash=stash, extra_stop_points=base_extra_stop_points | self._extra_stop_points, **kwargs)
  File "/root/angr-dev/angr/angr/manager.py", line 604, in step
    pg = pg._one_step(stash=stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/manager.py", line 352, in _one_step
    out = hook(pg, stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/exploration_techniques/looplimiter.py", line 17, in step
    pg = pg.step(stash=stash, **kwargs).move(stash, self.discard_stash,
  File "/root/angr-dev/angr/angr/manager.py", line 604, in step
    pg = pg._one_step(stash=stash, selector_func=selector_func, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/manager.py", line 381, in _one_step
    result_stashes = self._one_state_step(a, successor_func=successor_func, **kwargs)
  File "/root/angr-dev/angr/angr/manager.py", line 259, in _one_state_step
    ss = self._project.factory.successors(a, **kwargs)
  File "/root/angr-dev/angr/angr/factory.py", line 77, in successors
    r = engine.process(state, inline=inline,**kwargs)
  File "/root/angr-dev/angr/angr/engines/vex/engine.py", line 101, in process
    opt_level=opt_level)
  File "/root/angr-dev/angr/angr/engines/engine.py", line 54, in process
    self._process(new_state, successors, *args, **kwargs)
  File "/root/angr-dev/angr/angr/engines/vex/engine.py", line 151, in _process
    self._handle_irsb(state, successors, irsb, skip_stmts, last_stmt, whitelist)
  File "/root/angr-dev/angr/angr/engines/vex/engine.py", line 223, in _handle_irsb
    self._handle_statement(state, successors, stmt)
  File "/root/angr-dev/angr/angr/engines/vex/engine.py", line 343, in _handle_statement
    exit_stmt_idx=state.scratch.stmt_idx, exit_ins_addr=state.scratch.ins_addr)
  File "/root/angr-dev/angr/angr/engines/successors.py", line 116, in add_successor
    self._preprocess_successor(state, add_guard=add_guard)
  File "/root/angr-dev/angr/angr/engines/successors.py", line 146, in _preprocess_successor
    state.add_constraints(state.scratch.guard)
  File "/root/angr-dev/angr/angr/sim_state.py", line 336, in add_constraints
    added = self.se.add(*constraints)
  File "/root/angr-dev/angr/angr/state_plugins/sim_action_object.py", line 53, in ast_stripper
    return f(*new_args, **new_kwargs)
  File "/root/angr-dev/angr/angr/state_plugins/solver.py", line 88, in wrapped_f
    return f(*args, **kwargs)
  File "/root/angr-dev/angr/angr/state_plugins/solver.py", line 464, in add
    return self._solver.add(cc)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/constraint_fixer_mixin.py", line 9, in add
    return super(ConstraintFixerMixin, self).add(constraints, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/constraint_filter_mixin.py", line 27, in add
    return super(ConstraintFilterMixin, self).add(ec, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/constraint_deduplicator_mixin.py", line 38, in add
    added = super(ConstraintDeduplicatorMixin, self).add(filtered, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/sat_cache_mixin.py", line 26, in add
    added = super(SatCacheMixin, self).add(constraints, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/simplify_skipper_mixin.py", line 26, in add
    added = super(SimplifySkipperMixin, self).add(*args, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontends/composite_frontend.py", line 241, in add
    child_added += self._add_dependent_constraints(names, set_constraints, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontends/composite_frontend.py", line 223, in _add_dependent_constraints
    added = s.add(constraints, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/constraint_deduplicator_mixin.py", line 38, in add
    added = super(ConstraintDeduplicatorMixin, self).add(filtered, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/sat_cache_mixin.py", line 26, in add
    added = super(SatCacheMixin, self).add(constraints, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/simplify_skipper_mixin.py", line 26, in add
    added = super(SimplifySkipperMixin, self).add(*args, **kwargs)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 149, in add
    still_valid = set(self._get_models(extra_constraints=added))
  File "/root/angr-dev/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 208, in _get_models
    if m.eval_constraints(extra_constraints):
  File "/root/angr-dev/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 66, in eval_constraints
    return all(self.eval_ast(c) for c in constraints)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 66, in <genexpr>
    return all(self.eval_ast(c) for c in constraints)
  File "/root/angr-dev/claripy/claripy/frontend_mixins/model_cache_mixin.py", line 57, in eval_ast
    new_ast = ast._replace(self.replacements, leaf_operation=self._leaf_op)
  File "/root/angr-dev/claripy/claripy/ast/base.py", line 616, in _replace
    new_a = a._replace(replacements=replacements, variable_set=variable_set, leaf_operation=leaf_operation)
  File "/root/angr-dev/claripy/claripy/ast/base.py", line 616, in _replace
    new_a = a._replace(replacements=replacements, variable_set=variable_set, leaf_operation=leaf_operation)
  File "/root/angr-dev/claripy/claripy/ast/base.py", line 624, in _replace
    r = self.make_like(self.op, tuple(new_args))
  File "/root/angr-dev/claripy/claripy/ast/bits.py", line 20, in make_like
    return Base.make_like(self, *args, **kwargs)
  File "/root/angr-dev/claripy/claripy/ast/base.py", line 306, in make_like
    return type(self)(*args, **kwargs)
  File "/root/angr-dev/claripy/claripy/ast/base.py", line 134, in __new__
    r = operations._handle_annotations(eb._abstract(eb.call(op, args)), args)
  File "/root/angr-dev/claripy/claripy/backends/__init__.py", line 199, in call
    obj = self._op_raw[op](*converted)
  File "/root/angr-dev/claripy/claripy/bv.py", line 40, in normalize_helper
    return f(self, o)
  File "/root/angr-dev/claripy/claripy/bv.py", line 15, in compare_guard
    return f(self, o)
  File "/root/angr-dev/claripy/claripy/bv.py", line 397, in SMod
    division_result = a//b if a*b>0 else (a+(-a%b))//b
ZeroDivisionError: long division or modulo by zero
```

**Comment by ltfish** [comment]
This ZeroDivisionError should be caught somewhere in the call stack and raise a subclass of ClaripyError instead. My gut feeling is that we should catch it in `SMod`. @zardus What do you think?

**Comment by zardus** [comment]
Yes, similar to the other division operators in bv.py.

**Comment by rhelmot** [comment]
this was fixed in 8ea323cb
